### PR TITLE
Modify the differentiable HF solver function names

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
This PR updates the functions names in `qml.hf` following the discussion [here](https://docs.google.com/document/d/19B0M0KSoLppzjVopHJJeoez6uDCA4Qzfqrch7vcXf0Q/edit).

**Description of the Change:**
The term _generate_ is removed from the integrals and matrices function names.

**Benefits:**
The new names are more expressive and user friendly.

**Possible Drawbacks:**
The demos and other docs need to be updated.

**Related GitHub Issues:**
NA

### CODE REVIEW

File: pennylane/hf/hamiltonian.py [hunk 1]
```diff
@@ -172,14 +172,14 @@ def fermionic_hamiltonian(*args):
...
+        core_constant, one, two = electron_integrals(mol, core, active)(*args)
 
         return fermionic_observable(core_constant, one, two, cutoff)
 
-    return fermionic_hamiltonian
+    return _fermionic_hamiltonian
 
 
-def generate_hamiltonian(mol, cutoff=1.0e-12, core=None, active=None):
+def molecular_hamiltonian(mol, cutoff=1.0e-12, core=None, active=None):
```
Comment (soranjh) [inline pennylane/hf/hamiltonian.py:182]: This has a naming conflict with `qml.qchem.molecular_hamiltonian`.

Comment (ixfoduap) [reply pennylane/hf/hamiltonian.py:182]: I think I like `mol_hamiltonian`, which leads to nice syntax `mol_hamiltonian(mol)`. But curious to hear other thoughts

**Review by ixfoduap — CHANGES_REQUESTED** [review]

File: pennylane/hf/hamiltonian.py (hunk 1 above)
Comment (soranjh) [reply pennylane/hf/hamiltonian.py:182]: `mol_hamiltonian` is shorter than `molecular_hamiltonian` and I am fine with it.

File: pennylane/hf/__init__.py [hunk 2]
```diff
@@ -18,36 +18,32 @@
...
-    generate_repulsion_tensor,
-    molecular_density_matrix,
+    attraction_matrix,
+    core_matrix,
+    kinetic_matrix,
+    mol_density_matrix,
     moment_matrix,
+    overlap_matrix,
+    repulsion_tensor,
 )
```
Comment (josh146) [inline pennylane/hf/__init__.py:47]: @soranjh I noticed that there is still `generate_symmetries` present in the documentation?

File: pennylane/hf/dipole.py [hunk 3]
```diff
@@ -103,7 +103,7 @@ def _dipole_integrals(*args):
             tuple[array[float]]: tuple containing the core orbital contributions and the dipole
             moment integrals
         """
-        _, coeffs, _, _, _ = qml.hf.generate_scf(mol)(*args)
+        _, coeffs, _, _, _ = qml.hf.scf(mol)(*args)
```
Comment (josh146) [inline pennylane/hf/dipole.py:106]: very nice!

File: pennylane/hf/hamiltonian.py (hunk 1 above)
Comment (josh146) [reply pennylane/hf/hamiltonian.py:182]: I have no strong opinions either way :+1:

You have two options:

- Go through the docs and demos and change every `:func:~.molecular_hamiltonian` to either `:func:~.hf.molecular_hamiltonian` or `:func:~.qchem.molecular_hamiltonian`

- Change one of the function names.

Probably easier to do the latter 🙂

**Review by josh146 — COMMENTED** [review]
Don't forget to also update the `doc/introduction/chemistry.rst` file @soranjh!

I think you have also already seen the warnings in the doc CI?

https://github.com/PennyLaneAI/pennylane/runs/5410631507?check_suite_focus=true#step:4:2185

Aside from that, nothing blocking on my side :)

**Review by obliviateandsurrender — COMMENTED** [review]
Looking good @soranjh.

File: pennylane/hf/__init__.py (hunk 2 above)
Comment (soranjh) [reply pennylane/hf/__init__.py:47]: Hi Josh. The tapering functions require: renaming, some modification in format and re-placing based on our discussion [here](https://docs.google.com/document/d/19B0M0KSoLppzjVopHJJeoez6uDCA4Qzfqrch7vcXf0Q/edit). We will do these changes in a separate PR dedicated to tapering functionality.

**Review by ixfoduap — APPROVED** [review]

**Review by obliviateandsurrender — APPROVED** [review]

**Pull request merged**